How ClickHouse helps us track billions of GraphQL requests monthly (#893)
the-guild-org/the-guild-website: website/pages/blog/graphql-hive-and-clickhouse.mdx

---
title: How ClickHouse helps us track billions of GraphQL requests monthly
tags: [graphql, graphql-hive]
authors: [kamil]
date: 2022-11-02
description: Learn how ClickHouse enabled GraphQL Hive to scale from millions to billions of requests monthly.
image: /blog-assets/announcing-graphql-hive/cover.png
thumbnail: /blog-assets/announcing-graphql-hive/thumbnail.png
---

> You will learn how **ClickHouse** and **Kafka** enabled GraphQL Hive to collect billions of requests monthly without a drop of sweat.
> We will also cover the challenges we faced and how we overcame them. The article explains the architecture of our system in detail.

Reviewer: ```suggestion
You will learn how **ClickHouse** and **Kafka** enabled GraphQL Hive to collect billions of requests monthly without a drop of sweat.
We will also cover the challenges we faced and how we overcame them by covering Hive's architecture in detail.
```

      "operation": "query users { users { id } }",
      "fields": ["Query", "Query.users", "User", "User.id"]
    }
    /* ... */
  },
  "operations": [
    {
      "operationMapKey": "762a45e3", // points to the 'users' query
      "timestamp": 1663158676589,
      "execution": {
        "ok": true,
        "duration": 150000000, // 150ms in nanoseconds
        "errorsTotal": 0
      },
      "metadata": {
        "client": {
          "name": "demo", // Name of the GraphQL API consumer
          "version": "0.0.1" // Version of the GraphQL API consumer
        }
      }
    }
    /* ... */
  ]
}
```

## Why ClickHouse?

The kind of data GraphQL Hive deals with is not meant to be mutated (it never changes) and the database has to support a good set of functions for analytics and data aggregation.

Our first choice was **ElasticSearch**. It seemed like a good option, but we hit the first scaling issues very quickly.
**After reaching the goal of 100M operations per month, the queries to ElasticSearch were unusable.**
We tried to improve it, but we weren't happy with ElasticSearch anyway.

After looking for alternatives we decided to switch to ClickHouse.

> We highly recommend reading the "Migration from ElasticSearch to ClickHouse" article we wrote with the [ClickHouse Cloud](https://clickhouse.com/cloud) team.

Author: ```suggestion
We highly recommend reading the ["100x Faster: GraphQL Hive migration from Elasticsearch to ClickHouse"](https://clickhouse.com/blog/100x-faster-graphql-hive-migration-from-elasticsearch-to-clickhouse) article we wrote with the [ClickHouse Cloud](https://clickhouse.com/cloud) team.
```
